Delay serialization of Response

banacorn opened the pull request:
https://github.com/agda/agda/blob/fd3628d960ba3982022531aa582885752cba7930/src/full/Agda/Interaction/Response.hs#L29-L34

The comment says that `Response` is for any "interactive interface", but I found it very difficult for the interactive interfaces to make use of Response. 

https://github.com/agda/agda/blob/fd3628d960ba3982022531aa582885752cba7930/src/full/Agda/Interaction/Response.hs#L73-L77

Because everything was converted to `String`s a bit too early, before we get to choose which interactive interface to target.

I'm planning to recover the datatypes in `Response` and delay the process of serialization to a later stage (e.g. `lispifyResponse :: Response -> TCM [Lisp String]`). Please stop me if you think this is a bad idea.

- [x] Recover `Response`
  - [x]  Recover `DisplayInfo` 
    - [x]  Recover `Info_CompilationOk`
    - [x]  Recover `Info_Constraints`
    - [x]  Recover `Info_AllGoalsWarnings`
    - [x]  Recover `Info_Time`
    - [x]  Recover `Info_Error`
    - [x]  Recover `Info_Intro`
    - <del> Recover `Info_Auto` </del>
    - [x]  Recover `Info_ModuleContents`
    - [x]  Recover `Info_SearchAbout`
    - [x]  Recover `Info_WhyInScope`
    - [x]  Recover `Info_NormalForm`
    - [x]  Recover `Info_GoalType`
    - [x]  Recover `Info_CurrentGoal`
    - [x]  Recover `Info_InferredType`
    - [x]  Recover `Info_Context`
    - [x]  Recover `Info_HelperFunction`
  - <del>  Recover `Resp_GiveAction` </del>
  - <del>  Recover `Resp_MakeCase` </del>
 
EDIT: `Resp_GiveAction`, `Resp_MakeCase` and `Info_Auto` will not be covered in this PR. They deserve another PR of their own. 

Related issues: #3414, #3221 

## Possible future work/proposal 

### 1. Lift the serialization to `CommandM`?

Serialization mostly happens in `TCM`, however, some commands like `Cmd_infer_toplevel` are even more demanding, which requires the state from `CommandM`. The current workaround it to smuggle the state out from `CommandM`, and set up a new `CommandM` with that state when serializing.
I think I should do away with the workaround and instead just serialize everything in `CommandM`.

### 2. Construct a read-only `CommandM` stack for serialization?

Serialization should be read-only and should not modify any state.
It would be nice if we could impose that by replacing `CommandM` with some reader monad.
But I will only do this if most of the state-access functions can be reused and the codebase won't bloat.

banacorn committed:
[ refactor ] Add type synonym OpenMetas

Put visible and hidden metas together for later use.

banacorn committed:
[ refactor ] Delay serialization of open metas

`Info_AllGoalsWarnings` now carries `OpenMetas` instead of `String`.
So that we get to choose to target Emacs or JSON at the later stage.

banacorn committed:
[ refactor ] Renaming OpenMetas to Goals

To avoid confusion with `OpenMetas` in Agda.TypeChecking.MetaVars

banacorn committed:
[ refactor ] Delay serialization of TCWarnings

banacorn committed:
[ refactor ] Restore Info_Constraints

banacorn committed:
[ refactor ] Recover Info_CompilationOk

banacorn committed:
[ refactor ] Recover Info_Time

banacorn committed:
[ fix ] Fix and codify Warnings & Non-fatal Errors

banacorn committed:
[ refactor ] Modeling variants of Info_Error

banacorn committed:
[ refactor ] Recover Info_Error

banacorn committed:
[ refactor ] Attempting to restore Info_Intro

banacorn committed:
[ refactor ] Splitting constructors of Info_Intro

banacorn committed:
[ refactor ] Recover Info_ModuleContents

banacorn committed:
[ refactor ] Recover Intro_SearchAbout

banacorn committed:
[ refactor ] Recover Intro_WhyInScope

banacorn commented:
I'm marking this as ready for review because the CIs are not checking draft PRs.

banacorn committed:
[ refactor ] Splitting constructor Info_NormalForm

banacorn committed:
[ refactor ] Splitting InteractionTop.hs

banacorn committed:
[ refactor ] Recover Intro_NormalForm_TopLevel

banacorn committed:
[ fix] Adding Interaction/Base.hs to Agda.cabal

banacorn committed:
[ refactor ] Split & recover Info_InferredType

banacorn committed:
[ refactor ] Recover Info_Context

banacorn committed:
[ refactor ] Recover Info_GoalType

banacorn committed:
[ refactor ] Recover Info_CurrentGoal

banacorn committed:
[ refactor ] Recover Info_HelperFunction

banacorn committed:
[ refactor ] Cleanup and remove redundant imports

banacorn committed:
[ fix ] Fix the context of Intro_HelperFunction

banacorn committed:
[ refactor ] Test shielding local TCState

Let `localTCState` come after every `B.withInteractionId ii`.
We can further refactor this combo if it succeed.

banacorn committed:
[ refactor ] Split and refactor DisplayInfo

Split DisplayInfo into top-level/goal-specific variants.
Handle contexts of goal-specific uniformly.

banacorn committed:
[ refactor ] Rename constructors of DisplayInfo

banacorn committed:
[ refactor ] Relocating Goals & WarningsAndNonFatalErrors

banacorn committed:
[ refactor ] Remove dead code

gallais commented:
Sorry, I completely forgot about this. I will have a look today.

banacorn commented:
I almost forgot about this, too!

gallais commented on src/full/Agda/Interaction/BasicOps.hs:
```diff
@@ -577,6 +581,47 @@ getConstraints' g f = liftTCM $ do
+   where isNotMeta w = case tcWarning w of
+                         UnsolvedInteractionMetas{} -> False
+                         UnsolvedMetaVariables{}    -> False
+                         _                          -> True
```
We could define `isMetaWarning` in `Agda.TypeChecking.Warnings` just like
we already have `isUnsolvedWarning` and use that instead rather than an
ad-hoc auxiliary function.

gallais commented on src/full/Agda/Interaction/BasicOps.hs:
```diff
@@ -577,6 +581,47 @@ getConstraints' g f = liftTCM $ do
+getWarningsAndNonFatalErrors = do
+  mws <- getMaybeWarnings AllWarnings
+  return $ case filter isNotMeta <$> mws of
+    SomeWarnings ws@(_:_) -> swap $ classifyWarnings ws
```
The fact this pair has type `([a],[a])` and that we sometimes mean to return it
one way and sometimes the other suggests to me that we should introduce a record
type for it.

We could move `WarningsAndNonFatalErrors` to `Agda.TypeChecking.Warnings` and
change its definition from a mere type alias to:

```haskell
data WarningsAndNonFatalErrors = WarningsAndNonFatalErrors
  { warnings       :: [TCWarning]
  , nonFatalErrors :: [TCWarning]
  }
```

and not export the constructor to force us to always use the projections...

banacorn replied:
I'm still exporting the constructor or `WarningsAndNonFatalErrors`, because `getWarningsAndNonFatalErrors` in `Agda.TypeChecking.BasicOps` needs it.

Or should I:
1. move `getWarningsAndNonFatalErrors` to `Agda.TypeChecking.Warnings`?
2. define and export `WarningsAndNonFatalErrors [] []` as `emptyWarningsAndNonFatalErrors`?

```haskell
getWarningsAndNonFatalErrors :: TCM WarningsAndNonFatalErrors
getWarningsAndNonFatalErrors = do
  mws <- getMaybeWarnings AllWarnings
  let notMetaWarnings = filter (not . isMetaWarning) <$> mws
  return $ case notMetaWarnings of
    SomeWarnings ws@(_:_) -> classifyWarnings ws
    _ -> WarningsAndNonFatalErrors [] []
```

gallais replied:
2. is probably a lot simpler (`getMaybeWarnings` is defined in
`Interaction.Imports` so `getWarningsAndNonFatalErrors` would create
yet another dependency cycle)

By the way, `isMetaWarning` looks more like `isMetaTCWarning` to me.
I think it's worth defining the `isMetaWarning` and then filter using
`not . isMetaWarning . tcWarning`.

gallais commented on src/full/Agda/Interaction/InteractionTop.hs:
```diff
@@ -832,38 +489,30 @@ interpret :: Interaction -> CommandM ()
+  -- unsolvedNotOK <- lift $ not . optAllowUnsolved <$> pragmaOptions
+  ms <- lift B.getGoals
+  -- (we, wa) <- lift getWarningsAndNonFatalErrors
+  display_info . Info_AllGoalsWarnings ms =<< lift getWarningsAndNonFatalErrors
```
These lines which have been commented out can be removed.

gallais commented:
Cool stuff!

banacorn committed:
[ refactor ] Define `isMetaWarning`

Define `isMetaWarning` in `Agda.TypeChecking.Warnings` as
suggested by @gallais

banacorn committed:
[ refactor ] Codify `WarningsAndNonFatalErrors`

Define `WarningsAndNonFatalErrors` in `Agda.TypeChecking.Warnings`
and avoid import cycles.

banacorn committed:
[ refactor ] Hide constructor of `WarningsAndNonFatalErrors`

Export only the type and projections of `WarningsAndNonFatalErrors`.

banacorn committed:
[ refactor ] Rename and define `isMetaTCWarning`

andreasabel commented on src/full/Agda/Interaction/InteractionTop.hs:
```diff
@@ -1080,46 +731,47 @@ interpret (Cmd_elaborate_give norm ii rng s) = do
-    return $ "Elaborates to:" <+> txt
-  cmd_goal_type_context_and have norm ii rng s
+    normalForm norm term
+  cmd_goal_type_context_and (GoalAndElaboration term) norm ii rng s
```
This might have introduced regression #7624
- #7624
